### PR TITLE
Functions schema integration

### PR DESCRIPTION
**Pull request by mgodave**
Unify SerDe<T> and Schema<T>. Allow SerDe/Schema to be used when creating PulsarSink and PulsarSource.

### CODE REVIEW

**Comment by sijie** [comment]
@srkukarni @jerrypeng please review

File: pulsar-client/src/main/java/org/apache/pulsar/client/api/ConsumerBuilder.java [hunk 1]
```diff
@@ -94,6 +94,11 @@
      */
     ConsumerBuilder<T> topics(List<String> topicNames);
 
+    /**
+     *
+     */
+    ConsumerBuilder<T> addTopic(String topicName, Schema<T> schema);
```
Comment (merlimat) [inline pulsar-client/src/main/java/org/apache/pulsar/client/api/ConsumerBuilder.java]: This is a bit different from the the other `topic()` `topics()` methods. 

Also I find it confusing having a per topic schema, when the consumer itself needs to be created with a schema

**Comment by srkukarni** [comment]
Most of the changes in this pr touch producer/consumer. Plus there are a bunch of other changes necessary for functions that are not there yet(like schema validation at client/server side). My suggestion would to limit this change to the core pulsar changes and follow up with a different and more comprehensive pr that deals with just functions

**Comment by jerrypeng** [comment]
Yup I agree with @srkukarni

File: pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/source/PulsarSource.java [hunk 2]
```diff
@@ -60,12 +78,17 @@ public void open(Map<String, Object> config) throws Exception {
...
-                .topics(new ArrayList<>(this.pulsarSourceConfig.getTopicSerdeClassNameMap().keySet()))
+        ConsumerBuilder<T> consumerBuilder =
+            this.pulsarClient.newConsumer(emptySchema)
                 .subscriptionName(this.pulsarSourceConfig.getSubscriptionName())
                 .subscriptionType(this.pulsarSourceConfig.getSubscriptionType().get())
-                .ackTimeout(1, TimeUnit.MINUTES)
-                .subscribe();
+                .ackTimeout(1, TimeUnit.MINUTES);
+
+        topicToSerDeMap.forEach(consumerBuilder::addTopic);
```
Comment (jerrypeng) [inline pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/source/PulsarSource.java]: topicToSerDeMap is never used and consumer is NOT passed in any topics.

File: pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/source/PulsarSource.java [hunk 3]
```diff
@@ -34,17 +40,29 @@
...
-
 @Slf4j
 public class PulsarSource<T> implements Source<T> {
 
     private PulsarClient pulsarClient;
     private PulsarSourceConfig pulsarSourceConfig;
-    private Map<String, SerDe> topicToSerDeMap = new HashMap<>();
+    private Map<String, SerDe<T>> topicToSerDeMap = new HashMap<>();
+
+    private final Schema<T> emptySchema = new Schema<T>() {
```
Comment (jerrypeng) [inline pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/source/PulsarSource.java]: Why is this needed?

Comment (mgodave) [reply pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/source/PulsarSource.java]: Because of the aforementioned issue raised by @merlimat; a Consumer must be created with a schema so I needed something to capture the type before I added the individual topic/schema pairs.

File: pulsar-client/src/main/java/org/apache/pulsar/client/api/ConsumerBuilder.java (hunk 1 above)
Comment (merlimat) [reply pulsar-client/src/main/java/org/apache/pulsar/client/api/ConsumerBuilder.java]: > Multi-subscribe assumes that everything has a single schema (this may be a flawed assumption) 

I think that's an intuitive assumption: 
 * Consumer has a type 
 * It can consume from multiple topics.. as long as data has same format

To consume from different topics, with different types, we should just use different consumer instances.

**Comment by sijie** [comment]
what is the take here? @jerrypeng @mgodave @merlimat

**Comment by mgodave** [comment]
I'll rework this after the blog post.

File: pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/source/PulsarSource.java (hunk 3 above)
Comment (jerrypeng) [reply pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/source/PulsarSource.java]: is there a way for us the avoid this?  This can be confusing for the reader.

**Comment by mgodave** [comment]
I'm reworking it. Hopefully we won't need it.

On Wed, May 30, 2018 at 2:54 PM, Boyang Jerry Peng <notifications@github.com

**Comment by mgodave** [comment]
Please take another look @jerrypeng and @merlimat

**Comment by mgodave** [comment]
retest this please

**Comment by mgodave** [comment]
retest this please

File: pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/instance/JavaInstance.java [hunk 4]
```diff
@@ -49,14 +54,11 @@ public JavaInstance(InstanceConfig config, Object userClassObject,
                  PulsarClient pulsarClient,
                  Source source) {
         // TODO: cache logger instances by functions?
-        Logger instanceLog = LoggerFactory.getLogger("function-" + config.getFunctionDetails().getName());
-
-        if (source instanceof PulsarSource) {
```
Comment (jerrypeng) [inline pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/instance/JavaInstance.java]: Why did we get rid of this logic? We need to set ContextImpl accordingly

Comment (mgodave) [reply pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/instance/JavaInstance.java]: Look a little closer. it's now under handleMessage. We're building N consumers now and we need to set the appropriate Consumer in the context.

File: pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/instance/JavaInstance.java [hunk 5]
```diff
@@ -67,6 +69,13 @@ public JavaInstance(InstanceConfig config, Object userClassObject,
     }
 
     public JavaExecutionResult handleMessage(MessageId messageId, String topicName, Object input) {
+        if (source instanceof PulsarSource) {
```
Comment (srkukarni) [inline pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/instance/JavaInstance.java]: This is actually instantiating a context object per message. I think that should be avoided.

File: pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/source/PulsarSource.java [hunk 6]
```diff
@@ -108,31 +124,52 @@ public void open(Map<String, Object> config) throws Exception {
...
+            }).failFunction(() -> {
+                if (pulsarSourceConfig.getProcessingGuarantees() == FunctionConfig.ProcessingGuarantees.EFFECTIVELY_ONCE) {
+                    throw new RuntimeException("Failed to process message: " + message.getMessageId());
+                }
+            })
+            .build();
+
+        try {
+            recordsQueue.put(pulsarMessage);
+        } catch (InterruptedException e) {
```
Comment (srkukarni) [inline pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/source/PulsarSource.java]: We need to log this. Also is ignoring the right behaviour?

File: pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/instance/JavaInstance.java (hunk 4 above)
Comment (jerrypeng) [reply pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/instance/JavaInstance.java]: We should try to create  a new ContextImpl object in handle message since that is called on every message.  This may incur high GC costs

File: pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/source/PulsarSource.java [hunk 7]
```diff
@@ -36,20 +53,16 @@
 import org.apache.pulsar.io.core.Source;
 import org.jboss.util.Classes;
 
-import java.util.ArrayList;
-import java.util.HashMap;
-import java.util.Map;
-import java.util.concurrent.TimeUnit;
-
 @Slf4j
-public class PulsarSource<T> implements Source<T> {
```
Comment (jerrypeng) [inline pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/source/PulsarSource.java]: since we are pretty much make the Pulsar source into a push source here, can we just use the following push source abstract class:

https://github.com/apache/incubator-pulsar/blob/master/pulsar-io/core/src/main/java/org/apache/pulsar/io/core/PushSource.java

Comment (mgodave) [reply pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/source/PulsarSource.java]: this looked completely different before I merged master...

**Comment by sijie** [comment]
retest this please

File: pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/instance/JavaInstance.java [hunk 8]
```diff
@@ -38,22 +41,24 @@
...
     private java.util.function.Function javaUtilFunction;
+    private Optional<PulsarSource> optionalPulsarSource = Optional.empty();
 
     public JavaInstance(InstanceConfig config, Object userClassObject,
                  ClassLoader clsLoader,
                  PulsarClient pulsarClient,
                  Source source) {
         // TODO: cache logger instances by functions?
-        Logger instanceLog = LoggerFactory.getLogger("function-" + config.getFunctionDetails().getName());
+        this.instanceLog = LoggerFactory.getLogger("function-" + config.getFunctionDetails().getName());
```
Comment (jerrypeng) [inline pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/instance/JavaInstance.java]: I don't think we need a instanceLog to be a member variable.  I think it is only referenced in the constructor

File: pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/source/PulsarSource.java [hunk 9]
```diff
@@ -64,28 +66,32 @@ public void open(Map<String, Object> config) throws Exception {
...
 
-        // Setup pulsar consumer
-        ConsumerBuilder<byte[]> consumerBuilder = this.pulsarClient.newConsumer()
+        inputConsumers = Maps.newHashMap();
+        for (Map.Entry<String, SerDe<T>> entry : topicToSerDeMap.entrySet()) {
+            ConsumerBuilder<T> consumerBuilder = this.pulsarClient.newConsumer(entry.getValue())
                 .subscriptionName(this.pulsarSourceConfig.getSubscriptionName())
-                .subscriptionType(this.pulsarSourceConfig.getSubscriptionType());
+                .subscriptionType(this.pulsarSourceConfig.getSubscriptionType())
+                .ackTimeout(1, TimeUnit.MINUTES)
```
Comment (jerrypeng) [inline pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/source/PulsarSource.java]: This shouldn't be here

Comment (jerrypeng) [reply pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/source/PulsarSource.java]: if pulsarSourceConfig.getTimeoutMs() is null, we don't set the timeout

**Comment by sijie** [comment]
I merged the latest master. @jerrypeng also addressed your latest two comments (since they are simple comments to address)

**Comment by sijie** [comment]
retest this please

**Comment by sijie** [comment]
retest this please

**Comment by sijie** [comment]
(not sure why github checks hanging. trying close and reopen pull request to trigger CI again)

File: pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/instance/JavaInstance.java [hunk 10]
```diff
@@ -64,13 +68,17 @@ public JavaInstance(InstanceConfig config, Object userClassObject,
...
             this.javaUtilFunction = (java.util.function.Function) userClassObject;
         }
+
     }
 
     public JavaExecutionResult handleMessage(MessageId messageId, String topicName, Object input) {
-        if (context != null) {
-            context.setCurrentMessageContext(messageId, topicName);
-        }
+        optionalPulsarSource.ifPresent((pulsarSource) -> {
```
Comment (srkukarni) [inline pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/instance/JavaInstance.java]: what was the case where the context could be null?

**Comment by srkukarni** [comment]
retest this please

**Comment by sijie** [comment]
@jerrypeng @srkukarni can you guys review this PR again?

**Comment by srkukarni** [comment]
lgtm!

**Comment by sijie** [comment]
retest this please

**Comment by sijie** [comment]
retest this please
